Type/Password.pm

```diff
@@ -54,26 +53,25 @@ sub OTP_VALUE {
 }
 
 sub compare {
-    my($proto, $encrypted, $incoming) = @_;
+    my($proto, $hashed, $clear_text) = @_;
+    return -1
+        unless defined($hashed);
+    # Incoming clear text is never allowed to match stored OTP value and instead must be verified
+    # via the OTP module.
     return -1
-        unless defined($encrypted);
+        if $hashed eq $proto->OTP_VALUE;
     return 1
-        unless defined($incoming);
-    my($salt) = substr($encrypted, 0, 2);
-    my($i) = length($encrypted) == $_CRYPT_VALID_LENGTH
-        ? crypt($incoming, $salt)
-        : _encrypt($incoming, $salt);
-    return $encrypted cmp $i;
+        unless defined($clear_text);
+    my($hti) = _to_hash_type_instance_or_die($hashed);
+    return $hti->compare($clear_text);
 }
 
 sub encrypt {
-    my(undef, $password) = @_;
-    my($salt) = '';
-    for (my($i) = 0; $i < 2; $i++) {
-        $salt .= $_SALT_CHARS[int(rand($_SALT_INDEX_MAX) + 0.5)];
-    };
-    return _encrypt($password, $salt);
-
+    my(undef, $clear_text, $type) = @_;
+    $type ||= $_CURRENT_HASH_TYPE;
+    b_die("unsupported hash type=$type")
+        unless grep($_ eq $type, @$_HASH_TYPES);
```

Comment:
It might be worth creating a `$_VALID_HASH_TYPES` to simplify this code. A grep increases cognitive load.

Reply:
Not sure I follow how you're suggesting this code be simplified. `$_HASH_TYPES` is the list of supported hash types.

Reply:
```pl
my($_VALID_HASH_TYPE) = {map(($_ => 1), @$_HASH_TYPES)};
```

Then:
```pl
    b_die("unsupported hash type=$type")
        unless $_VALID_HASH_TYPE->{$type};
```

This could alternatively be a function, but it's easier for me to "valid_hash_type" than a grep.